### ISSUE TITLE
Feature Request: Command History Tab/Widget

### ISSUE THREAD

**Issue by dhirschfeld** [jupyterlab/jupyterlab#1284]
One of the most useful features of spyder is its *searchable* command history pane which I end up using quite a lot - I keep around the last 10k lines Just In Case

It would be good if JupyterLab had something similar. Would be happy for it to be an extension if the extension mechanism is able to handle this usecase

**Comment by blink1073** [comment]
As in, all of the kernel inputs executed while in JupyterLab?

**Comment by dhirschfeld** [comment]
Yep, just a listing of all commands previously entered which you can search is very useful.

Being able to type the start of a command to filter the history works well except when the command you want was not the first line in a multi-line block.

At least in spyder, if you enter a multi-line block:
```
import some_module
result = some_long_calculation_i_dont_want_to_type_again
```
Then if you type `result` and press up to cycle through the history it won't pick up the above block because it wasn't on the first line. To find the command you can then search the history tab for `result = `

Another benefit is that it allows you to see several matches at once so you can select the correct variant if there are multiple matches without having to individually cycle through them.

If this were to be implemented in Jupyter you'd probably want separate histories for different languages.

**Comment by JasonMDSI** [comment]
Can I find a history of all codes I executed in cells of a Jupyterlab (even the cells eventually deleted before saving the lab), similar to the History tab in RStudio?
